esmvaltool/preprocessor/_volume_pp.py

```diff
@@ -50,6 +51,93 @@ def volume_slice(cube, z_min, z_max):
+    """
+    Generate a new cube with the volume averaged data.
+
+    The resultant cube is seeded with `src_cube` metadata and coordinates,
+    excluding any source coordinates that span the associated vertical
+    dimension. The `times` of interpolation are used along with the
+    associated source cube time coordinate metadata to add a new
+    time coordinate to the resultant cube.
+
+    Based on the _create_cube method from _regrid.py.
+
+    Parameters
+    ----------
+    src_cube : cube
+        The source cube that was vertically interpolated.
+    data : array
+        The payload resulting from interpolating the source cube
+        over the specified times.
+    times : array
+        The array of times.
+
+    Returns
+    -------
+    cube
+
+    .. note::
+
+        If there is only one level of interpolation, the resultant cube
+        will be collapsed over the associated vertical dimension, and a
+        scalar vertical coordinate will be added.
+
+    """
+    print(src_cube.coords)
+    print(len(data))
+    print(times)
+    # Get the source cube vertical coordinate and associated dimension.
+    src_times = src_cube.coord('time')
+    t_dim, = src_cube.coord_dims(src_times)
+
+    if data.shape[t_dim] != len(times):
+        emsg = ('Mismatch between data and times for data dimension {!r}, '
+                'got data shape {!r} with times shape {!r}.')
+        raise ValueError(emsg.format(t_dim, data.shape, times.shape))
+
+    # Construct the resultant cube with the interpolated data
+    # and the source cube metadata.
+    kwargs = deepcopy(src_cube.metadata)._asdict()
+    result = iris.cube.Cube(data, **kwargs)
+
+    # Add the appropriate coordinates to the cube, excluding
+    # any coordinates that span the z-dimension of interpolation.
+    for coord in src_cube.dim_coords:
+        [dim] = src_cube.coord_dims(coord)
+        if dim != t_dim:
+            result.add_dim_coord(coord.copy(), dim)
+
+    for coord in src_cube.aux_coords:
+        dims = src_cube.coord_dims(coord)
+        if t_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+
+    for coord in src_cube.derived_coords:
+        dims = src_cube.coord_dims(coord)
+        if t_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+
+    # Construct the new vertical coordinate for the interpolated
+    # z-dimension, using the associated source coordinate metadata.
+    kwargs = deepcopy(src_times._as_defn())._asdict()
+
+    try:
+        coord = iris.coords.DimCoord(times, **kwargs)
+        result.add_dim_coord(coord, t_dim)
+    except ValueError:
+        coord = iris.coords.AuxCoord(times, **kwargs)
+        result.add_aux_coord(coord, t_dim)
+
+    # Collapse the z-dimension for the scalar case.
+    if times.size == 1:
+        slicer = [slice(None)] * result.ndim
+        slicer[t_dim] = 0
+        result = result[tuple(slicer)]
+
+    return result
+
+
```

Comment:
I don't understand this function - where is the interpolation actually taking place? It looks to me that you are creating a ```result``` cube from a source cube ```src_cube``` by copying a whole bunch of coordinates and metadata, which BTW I don't think you should, just say ```result = src_cube``` and modify in place the data and coordinates you need to modify; if you don't want to modify in place, ```copy()``` should still not be used, just assign the variables directly, without kwargs

Reply:
To be completely honest @valeriupredoi , I don't fully understand this function either. After several days of attempting all sorts of things, I eventually found a function in ESMValTool that can create a valid cube from scratch. This function is almost a word for word copy of the `_create_cube` method from the `_regrid.py` preprocessor. 

This is the only way I've found to make an object that iris recognised as a cube. I'd be happy to implement an alternative, though I'd strongly prefer not to spend too much more time on this. Please advise.

Reply:
no worries @ledm - for the future work here is a type of function that puts all sorts of stuff into a new cube and is loved by ```iris```- ```def _put_in_cube(template_cube, cube_data, stat_name,...``` in ```_multimodel.py```. Don't bother too much just now, important is it works :grin: